DynamicActiveQuery.php

```diff
@@ -27,6 +28,45 @@ class DynamicActiveQuery extends ActiveQuery
      */
     private $dynamicColumn;
 
+    private $wrapped = false;
+
+    /**
+     * Convert index value to closure, that will get decoded dynamic attribute, in case if indexing attribute is dynamic
+     * @param callable|string $column
+     * @return $this
+     */
+    public function indexBy($column)
+    {
+        if (!$this->asArray) {
+            return parent::indexBy($column);
+        }
+
+        /** @var \yii\db\ActiveRecord $modelClass */
+        $modelClass = $this->modelClass;
+        $this->indexBy = function ($row) use ($column, $modelClass) {
+            if (isset($row[$column])) {
+                return $row[$column];
+            }
+
+            if (!method_exists($modelClass, 'dynamicColumn')) {
+                throw new Exception("Indexable column {$column} does not exist");
+            }
+
+            $dynamicColumn = $modelClass::dynamicColumn();
+            if (!isset($row[$dynamicColumn])) {
+                throw new Exception("Dynamic column {$dynamicColumn} does not exist");
+            }
+
+            $dynamicAttributes = DynamicActiveRecord::dynColDecode($row[$dynamicColumn]);
+            $value = $this->getDotNotatedValue($dynamicAttributes, $column);
+
+            // in case if dynamic column does not exist for this row - use PK
```

Comment:
this seems odd. the idea of dynamic columns is that records frequently do not have a certain field. imagine indexing by that field's value for some records and by pk for others. how can that make sense? imagine the dyn col contains ints and the pk is an int. 

this makes me wonder if it makes any sense to support idnexBy() at all for dyn cols.

when a record does not have a certain field in the dyn cols, this, to me, is like an [sql null](https://en.wikipedia.org/wiki/Null_%28SQL%29). now imagine using indexBy() on a real table column with lots of nulls in it. what should AR do? i need to think about this.